### ISSUE TITLE
NAN disappears, but 0 appears

### ISSUE THREAD

**Issue by ke-dev** [open-mmlab/mmsegmentation#1632]
Hello, developer, I modified the pixels of the annotated image, starting from 0 to 28 in the background，but the result of the validation set is very unsatisfactory, do you know the reason for this?
![80e0906c6ee7f95416c0379631f3a22](https://user-images.githubusercontent.com/59656387/171349830-9e8d3978-2e5c-4d4a-b53a-99a8f859628c.png)
Here are some of my configs
![8573a9c9e68821dc95c4d57e555469b](https://user-images.githubusercontent.com/59656387/171349508-09f0f388-d658-4fa9-a195-f1196790092c.png)
![f08afe4ce5017e5c233f7bbe29f176a](https://user-images.githubusercontent.com/59656387/171349742-6f7509da-36d3-4729-ad3e-de3b53199ad9.png)

look forward to you early reply.

**Comment by ke-dev** [comment]
I did some comparative experiments and found that it may still be my data problem. I did this. In the label grayscale image, one label corresponds to one pixel value, and the others are all 0. What is wrong with this modification?

**Comment by ke-dev** [comment]
I did some comparative tests and found that it may still be my data problem. I did this. One label in the annotation map corresponds to one pixel value, and the others are all 0. What is wrong with this modification?

**Comment by MeowZheng** [comment]
Could you provide more details about how modified the classes of annotation?

**Comment by ke-dev** [comment]
sorry for the late reply.
Of course you can, here is the code I use.
`_json = glob.glob('./*.json')
 
name_dict = {
                "__background__": "0",
                "jb": "1",
                "ak": "2",
                "bb": "3",
                "zw": "4",
                "hh": "5",
                "dbh": "6",
                "fcb": "7",
                "lf": "8",
                "yw": "9",
                "xk": "10",
                "fbz": "11",
                "akd": "12",
                "zlb": "13",
                "ywq": "14",
                "gh": "15",
                "jy": "16",
                "qj": "17",
                "sr": "18",
                "yq": "19",
                "wyb": "20",
                "qb": "21",
                "yy": "22",
                "ms": "23",
                "th": "24",
                "mx": "25",
                "sz": "26",
                "lw": "27",
                "jh": "28"
            }
for j in _json:
    with open(j, 'r') as f:
        r = f.read()
    r = json.loads(r)
    obj = r['shapes']
 
    img = np.zeros((2048,2048), dtype="uint8")
    
    for i in obj:
        name = i['label']
        po = i['points']
        n = name_dict[name]
        print(name, po, n)

        img = cv2.fillPoly(img, [np.array(po, dtype=np.int32)], int(n), 0, 0)

    path = '../../ade_01/ADEChallengeData2016/annotations/training/'
    cv2.imwrite(os.path.join(path , j[:-4]+'png'), img)`

**Comment by ke-dev** [comment]
> Could you provide more details about how modified the classes of annotation?

sorry for the late reply.
Of course, here is the code I use:
![image](https://user-images.githubusercontent.com/59656387/172037438-4ebc95e3-6aff-42a2-9ae1-d801053664d5.png)

**Comment by MeowZheng** [comment]
Base on the classes of your dataset, I found the zero label will be predicted from the model, so I suggest `reduce_zero_label=False` in your train_pipeline.

**Comment by ke-dev** [comment]
> Base on the classes of your dataset, I found the zero label will be predicted from the model, so I suggest `reduce_zero_label=False` in your train_pipeline.

I set the `reduce_zero_label=False` as you said，but the results are still disappointing.
![377319387cbec7c8ad6f4c6608e855f](https://user-images.githubusercontent.com/59656387/172574728-4887551e-99b3-4dee-8e9a-4bb05dba5e5b.png)

Here are more details about my dataset
![b720dd687cbcc6a83ee7b878bff8e39](https://user-images.githubusercontent.com/59656387/172579289-5c2cf39e-1010-41c3-a8ff-a2a5187c5086.png)
does the black block area of the image (left) affect the result?
this is a portion of the pixel value of the annotated image
![a92919f64d3de69bc37a667049593fd](https://user-images.githubusercontent.com/59656387/172580293-aded2500-8b0a-4474-9463-113a44212b4a.png)
look forward to your reply.

**Comment by MeowZheng** [comment]
I think you might check the `num_classes` in you model cfg and `reduce_zero_label `, as  you said 29 classes ([0,28]) from 'background', but from this pic
![image](https://user-images.githubusercontent.com/76149310/172646635-ef7ebd41-07e7-46c0-a8ee-935574a55b45.png)
and this pic
![image](https://user-images.githubusercontent.com/76149310/172646731-030fd084-a719-4613-a1bb-28924cacac2b.png)
The num of class is 28

My suggest if you have 29 classes including 'background', please set
`num_classes=29` and `reduce_zero_label=False`
if you have 28 classes without 'background', please set
`num_classes=28` and `reduce_zero_label=True`

**Comment by binging512** [comment]
> I think you might check the `num_classes` in you model cfg and `reduce_zero_label `, as you said **29** classes ([0,28]) from 'background', but from this pic ![image](https://user-images.githubusercontent.com/76149310/172646635-ef7ebd41-07e7-46c0-a8ee-935574a55b45.png) and this pic ![image](https://user-images.githubusercontent.com/76149310/172646731-030fd084-a719-4613-a1bb-28924cacac2b.png) The num of class is **28**
> 
> My suggest if you have 29 classes including 'background', please set `num_classes=29` and `reduce_zero_label=False` if you have 28 classes without 'background', please set `num_classes=28` and `reduce_zero_label=True`

I think your code are all correct, since there is no NAN appears. However, I think your dataset is too difficult for the network to learn. You can see all the pixels are predicted to the 0 label. So maybe there is severe long-tailed distribution problems or other annotation problems. You can analysis your dataset and calculate the proportion for each category, and balance the effort of each category with weight in nn.CrossEntropyLoss().

**Comment by ke-dev** [comment]
> I think you might check the `num_classes` in you model cfg and `reduce_zero_label `, as you said **29** classes ([0,28]) from 'background', but from this pic ![image](https://user-images.githubusercontent.com/76149310/172646635-ef7ebd41-07e7-46c0-a8ee-935574a55b45.png) and this pic ![image](https://user-images.githubusercontent.com/76149310/172646731-030fd084-a719-4613-a1bb-28924cacac2b.png) The num of class is **28**
> 
> My suggest if you have 29 classes including 'background', please set `num_classes=29` and `reduce_zero_label=False` if you have 28 classes without 'background', please set `num_classes=28` and `reduce_zero_label=True`

I tried both but the result is the same，maybe you are right @binging512. one of the train log is here. #1615

**Comment by Winnie202** [comment]
> > 我想你可能会检查`num_classes`你的模型 cfg 和`reduce_zero_label `，正如你所说的来自“背景”的**29**个类（[0,28]），但是从这张图片![图片](https://user-images.githubusercontent.com/76149310/172646635-ef7ebd41-07e7-46c0-a8ee-935574a55b45.png)和这张图片![图片](https://user-images.githubusercontent.com/76149310/172646731-030fd084-a719-4613-a1bb-28924cacac2b.png)来看，类的数量是**28**
> > 我的建议是如果你有 29 个课程包括“背景”，请设置`num_classes=29`，`reduce_zero_label=False`如果你有 28 个课程没有“背景”，请设置`num_classes=28`和`reduce_zero_label=True`
> 
> 我都试过了，但结果是一样的，也许你是对的@binging512. 火车日志之一在这里。#1615

Is your problem solved? I have the same problem as you, I count the background into a total of 24 categories, and you have the same structure

**Comment by ke-dev** [comment]
> > > 我想你可能会检查`num_classes`你的模型 cfg 和`reduce_zero_label `，正如你所说的来自“背景”的**29**个类（[0,28]），但是从这张图片![图片](https://user-images.githubusercontent.com/76149310/172646635-ef7ebd41-07e7-46c0-a8ee-935574a55b45.png)和这张图片![图片](https://user-images.githubusercontent.com/76149310/172646731-030fd084-a719-4613-a1bb-28924cacac2b.png)来看，类的数量是**28**
> > > 我的建议是如果你有 29 个课程包括“背景”，请设置`num_classes=29`，`reduce_zero_label=False`如果你有 28 个课程没有“背景”，请设置`num_classes=28`和`reduce_zero_label=True`
> > 
> > 
> > 我都试过了，但结果是一样的，也许你是对的@binging512. 火车日志之一在这里。#1615
> 
> Is your problem solved? I have the same problem as you, I count the background into a total of 24 categories, and you have the same structure

yes, check your dataset carefully if your config is ok, I got a error while transforming the dataset and I fixed it

**Comment by Aniwine** [comment]
I think you should check the img_norm_cfg of pipeline ,I fixed my problem with setting the mean value to 0 and std to 1 on my two-classes task, it works.
